frontend/src/app/main/data/workspace/pages.cljs

```diff
@@ -167,44 +167,52 @@
             base-name          (:name page)
             name               (cfh/generate-unique-name base-name unames :suffix-fn suffix-fn)
             objects            (update-vals (:objects page) #(dissoc % :use-for-thumbnail))
-
-            main-instances-ids (set (keep #(when (ctc/main-instance? (val %)) (key %)) objects))
-            ids-to-remove      (set (apply concat (map #(cfh/get-children-ids objects %) main-instances-ids)))
-
+            ;; id-remap keeps track of the correspondence between original object ids and their new ids during page duplication.
+            ;; This is essential for updating parent-child relationships: when an object is cloned and assigned a new id, any parent referencing it must update its :shapes vector to use the new id.
+            ;; id-remap ensures referential integrity in the duplicated page structure.
+            id-remap           (atom {})
+            ;; Clone a main instance and its descendants, updating id-remap
             add-component-copy
-            (fn [objs id shape]
+            (fn [objs shape parent-id]
               (let [component (ctkl/get-component fdata (:component-id shape))
                     [new-shape new-shapes]
                     (ctn/make-component-instance page
```

Comment:
I've just found that `make-component-instance` function has a bug. It always sets the `parent-id` of the root of the new copy the same as the `frame-id`. It works when the copy is inside a frame, but not if it's under a group. I think you cannot cause this in normal cases, because you cannot drag a component into a group, but it happens in the present case when duplicating a page. I'd add a `force-parent-id` parameter to `make-component-instance` instead of setting it after the call. The parameter could be passed into `ctst/clone-shape` function inside.